Source/WebCore/rendering/RenderBlock.cpp

```diff
@@ -2422,6 +2422,18 @@ void RenderBlock::computeBlockPreferredLogicalWidths(LayoutUnit& minLogicalWidth
     maxLogicalWidth = std::max(floatLeftWidth + floatRightWidth, maxLogicalWidth);
 }
 
+std::optional<LayoutRect> RenderBlock::computeVisibleRectInContainer(const LayoutRect& rect, const RenderLayerModelObject* container, VisibleRectContext context) const
+{
+    auto result = RenderBox::computeVisibleRectInContainer(rect, container, context);
+    RenderBlock* caretPainter = frame().selection().caretRendererWithoutUpdatingLayout();
+    bool isContentEditable = frame().selection().selection().hasEditableStyle();
+
+    if (result && caretPainter == this && (isContentEditable || settings().caretBrowsingEnabled()))
```

Comment:
This is going back towards the approach we were trying to avoid with earlier changes.

RenderBlock shouldn't have to know about caretPainter behavior here. We need to pass some state down with the repaint that originates from the caret painter.

Reply:
Isn't that what we already do on FrameSelection.cpp:1767 with the call to `repaintRectangle`? That by-passes the call to `computeRectForRepaint` and returns the adjusted rectangle as we do here, but maybe there is an additional clip occurring in `RenderObject::repaintUsingContainer`?

Reply:
I investigated further and it looks like the rect is getting clipped in:
```
LayoutRect RenderBox::clippedOverflowRect(const RenderLayerModelObject* repaintContainer, VisibleRectContext context) const
```
which results in the artifacts since we don't repaint outside of the result of `clippedOverflowRect` from what I can tell.

Would it be acceptable @smfr to override `clippedOverflowRect` inside `RenderBlock`? I don't think passing state from the call to `repaintRectangle` in `FrameSelection.cpp:1767` will work unless that state is stored in the RenderObject for subsequent calls to `clippedOverflowRect` but that seems like a much worse approach.

The diff would look something like this:

```
diff --git a/Source/WebCore/rendering/RenderBlock.cpp b/Source/WebCore/rendering/RenderBlock.cpp
index 59af469ebb81..995e136008b0 100644
--- a/Source/WebCore/rendering/RenderBlock.cpp
+++ b/Source/WebCore/rendering/RenderBlock.cpp
@@ -2422,14 +2422,14 @@ void RenderBlock::computeBlockPreferredLogicalWidths(LayoutUnit& minLogicalWidth
     maxLogicalWidth = std::max(floatLeftWidth + floatRightWidth, maxLogicalWidth);
 }
 
-std::optional<LayoutRect> RenderBlock::computeVisibleRectInContainer(const LayoutRect& rect, const RenderLayerModelObject* container, VisibleRectContext context) const
+LayoutRect RenderBlock::clippedOverflowRect(const RenderLayerModelObject* repaintContainer, VisibleRectContext context) const
 {
-    auto result = RenderBox::computeVisibleRectInContainer(rect, container, context);
+    auto result = RenderBox::clippedOverflowRect(repaintContainer, context);
     RenderBlock* caretPainter = frame().selection().caretRendererWithoutUpdatingLayout();
     bool isContentEditable = frame().selection().selection().hasEditableStyle();
 
-    if (result && caretPainter == this && (isContentEditable || settings().caretBrowsingEnabled()))
-        return frame().selection().caretAnimator().caretRepaintRectForLocalRect(*result);
+    if (caretPainter == this && (isContentEditable || settings().caretBrowsingEnabled()))
+        return frame().selection().caretAnimator().caretRepaintRectForLocalRect(result);
 
     return result;
 }
```

Reply:
Discussed offline and I will take a closer look to see what this occurs on quip and in text input fields